django/transformer/transform/fhir.py

```diff
@@ -119,16 +121,32 @@ def handle_array_attributes(attributes_in_array, row):
         val = row.get(attr.path)
         if not isinstance(val, list) or len(val) == 1:
             continue
-        if length != 1 and len(val) != length:
+        if length not in (1, len(val)):
             raise ValueError("mismatch in array lengths")
         length = len(val)
 
-    # Now we can build the array
+    attributes_in_nested_arrays = {path: attr for path, attr in attributes_in_array.items() if has_index(path)}
+
     array = []
-    for index in range(length):
-        element = build_fhir_object(row, attributes_in_array, index=index)
-        if element is not None and element != {}:
-            array.append(element)
+    if attributes_in_nested_arrays and all(
+        path in attributes_in_nested_arrays or has_single_value(attr, row)
+        for path, attr in attributes_in_array.items()
+    ):
+        # If we have a nested array and all
+        # values outside of it come from the primary table
+        array.append(build_fhir_object(row, attributes_in_array))
+
+    else:
+        for index in range(length):
+            if attributes_in_nested_arrays:
+                # If we have a nested array and some values outside of it come from
+                # a joined table, we want to expand the outside array
+                indexed_row = {k: get_element_in_array(v, index) for k, v in row.items()}
+                element = build_fhir_object(indexed_row, attributes_in_array)
+            else:
+                element = build_fhir_object(row, attributes_in_array, index=index)
```

Comment:
add a comment to explicit what's happening here: we're expanding the nested (or inner, inside) array